File: Wrapper/Controller.cpp
#include "Controller.h"

#include <string>

namespace CLI
{
	Controller::Controller()
		: ManagedObject(new Core::Controller())
	{
		System::Diagnostics::Debug::WriteLine("Creating a new Entity-wrapper object.");
	}

	void Controller::DetectCoreCount() {
		m_Instance->DetectCoreCount();
	}

	void Controller::MoveAllAppsToEfficiencyCores() {
		m_Instance->MoveAllAppsToEfficiencyCores();
	}

	void Controller::MoveAllAppsToSomeEfficiencyCores() {
		m_Instance->MoveAllAppsToSomeEfficiencyCores();
	}

	bool Controller::MoveAppToHybridCores(String^ path, int eCores, int pCores)
	{
		IntPtr hGlobal = Marshal::StringToHGlobalUni(path);
		const wchar_t* unmanagedPath = static_cast<const wchar_t*>(hGlobal.ToPointer());
		bool success = m_Instance->MoveAppToHybridCores(unmanagedPath, eCores, pCores);
		Marshal::FreeHGlobal(hGlobal);

		return success;
	}

	void Controller::ResetToDefaultCores() {
		m_Instance->ResetToDefaultCores();
	}

	void Controller::MoveAllAppsToHybridCores(int eCores, int pCores) {
		m_Instance->MoveAllAppsToHybridCores(eCores, pCores);
	}

	int Controller::TotalCoreCount() {
		return m_Instance->TotalCoreCount();
	}

	int Controller::EfficiencyCoreCount() {
		return m_Instance->EfficiencyCoreCount();
	}

	int Controller::PerformanceCoreCount() {
		return m_Instance->PerformanceCoreCount();
	}
}

File: EnergyPerformance/Services/Controller.cs
using System.Diagnostics;
using EnergyPerformance.Helpers;

namespace EnergyPerformance.Services;

public sealed class Controller
{
    private readonly PipeClient _pipeClient;
    
    public Controller(PipeClient pipeClient)
    {
        _pipeClient = pipeClient;
    }
    
    public void MoveAllAppsToEfficiencyCores()
    {
        var command = "MoveAllAppsToEfficiencyCores";
        _pipeClient.SendMessage(command);
    }

    public void MoveAllAppsToSomeEfficiencyCores()
    {
        var command = "MoveAllAppsToSomeEfficiencyCores";
        _pipeClient.SendMessage(command);
    }
    
    public bool MoveAppToHybridCores(string target, int eCores, int pCores) 
    {
        var command = $"MoveAppToHybridCores {target} {eCores} {pCores}";
        var response = _pipeClient.SendAndReceiveMessage(command);
        return response == "true";
    }

    public void MoveAllAppsToHybridCores(int eCores, int pCores)
    {
        var command = $"MoveAllAppsToHybridCores {eCores} {pCores}";
        _pipeClient.SendMessage(command);
    }

    public void ResetToDefaultCores()
    {
        var command = "ResetToDefaultCores";
        _pipeClient.SendMessage(command);
    }

    public void DetectCoreCount()
    {
        var command = "DetectCoreCount";
        _pipeClient.SendMessage(command);
    }

    public int TotalCoreCount()
    {
        var command = "TotalCoreCount";
        var response = _pipeClient.SendAndReceiveMessage(command);
        return int.Parse(response);
    }

    public int EfficiencyCoreCount()
    {
        var command = "EfficiencyCoreCount";
        var response = _pipeClient.SendAndReceiveMessage(command);
        return int.Parse(response);
    }

    public int PerformanceCoreCount()
    {
        var command = "PerformanceCoreCount";
        var response = _pipeClient.SendAndReceiveMessage(command);
        Debug.WriteLine(response);
        return int.Parse(response);
    }
}
